include/range/v3/view/remove.hpp

```diff
@@ -0,0 +1,94 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Andrey Diduh 2019
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_VIEW_REMOVE_HPP
+#define RANGES_V3_VIEW_REMOVE_HPP
+
+#include <type_traits>
+#include <utility>
+#include <meta/meta.hpp>
+#include <range/v3/utility/invoke.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/view/remove_if.hpp>
+#include <range/v3/range_fwd.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \addtogroup group-views
+        /// @{
+        namespace view
+        {
+            struct remove_fn
+            {
+            private:
+                friend view_access;
+
+                template<typename Value, typename Proj>
+                static auto bind(remove_fn remove, Value value, Proj proj)
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    make_pipeable(std::bind(remove, std::placeholders::_1,
+                                  std::move(value),
+                                  protect(std::move(proj))))
+                )
+
+                template<typename Value>
+                static auto bind(remove_fn remove, Value value)
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    make_pipeable(std::bind(remove, std::placeholders::_1,
+                                  std::move(value)))
+                )
+
+                template<class Value>
+                struct Pred
+                {
+                    Value value;
+
+                    template<class Any>
+                    bool operator()(const Any& any) const
+                    {
+                        return any == value;
+                    }
+                };
+
+            public:
+                template<typename Rng, typename Value, typename Proj>
+                using Constraint = meta::and_<
+                        MoveConstructible<Value>,
+                        EqualityComparable< Value, typename projected<iterator_t<Rng>, Proj>::reference >,
+                        remove_if_fn::Constraint<Rng, Pred<Value>, Proj>>;
+
+                template<typename Rng, typename Value, typename Proj = ident,
+                        CONCEPT_REQUIRES_(Constraint<Rng, Value, Proj>())>
+                RANGES_CXX14_CONSTEXPR
+                auto operator()(Rng&& rng, Value&& value, Proj proj = Proj{}) const
```

Comment:
So, value should be moved inside Predicate? We will not support storing value as reference?

I do understand that it is safer to have value "stashed" inside view, since view can be moved around. But this narrows `view::remove` use case to move-only types.

Reply:
> But this narrows `view::remove` use case to move-only types.

I don't understand. I am not suggesting you change anything about how `remove_view` stores the value, only how this function takes the argument. Take the argument by value and move it in instead of taking it by forwarding reference and forwarding it in. That is:

```c++
               template<typename Rng, typename Value, typename Proj = ident,
                    CONCEPT_REQUIRES_(Constraint<Rng, Value, Proj>())>
                RANGES_CXX14_CONSTEXPR
                auto operator()(Rng&& rng, Value value, Proj proj = Proj{}) const
                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
                (
                    remove_if(all(static_cast<Rng &&>(rng)),
                        pred<detail::decay_t<Value>>{std::move(value)},
                        std::move(proj))
                )
```

This has the added benefit of actually making your constraint correct. Right now, if `Value` is deduced to be an lvalue reference type, you are testing whether the reference is move-constructible. That's not what you want.